File: backend/game/gme_sudoku/utils/get_is_board_valid.py
def get_is_area_unique(board, start_column, start_row, ignore_empty):
    area_values = []
    for column_idx in range(start_column, 3):
        for row_idx in range(start_row, 3):
            cur_item = board[row_idx][column_idx]
            if cur_item == '' and ignore_empty:
                continue
            if cur_item in area_values:
                return False
            area_values.append(cur_item)
    return True


def get_is_row_unique(board, row_idx, ignore_empty):
    row_values = []
    for column_idx in range(0, 9):
        cur_item = board[row_idx][column_idx]
        if cur_item == '' and ignore_empty:
            continue
        if cur_item in row_values:
            return False
        row_values.append(cur_item)
    return True


def get_is_column_unique(board, column_idx, ignore_empty):
    row_values = []
    for row_idx in range(0, 9):
        cur_item = board[row_idx][column_idx]
        if cur_item == '' and ignore_empty:
            continue
        if cur_item in row_values:
            return False
        row_values.append(cur_item)
    return True


def get_is_board_valid(board, ignore_empty=False):
    for row_idx in range(0, 9, 3):
        for column_idx in range(0, 9, 3):
            if not get_is_area_unique(board, row_idx, column_idx, ignore_empty):
                return False

    for row_idx in range(0, 9):
        if not get_is_row_unique(board, row_idx, ignore_empty):
            return False

    for column_idx in range(0, 9):
        if not get_is_column_unique(board, column_idx, ignore_empty):
            return False

    return True

File: backend/src/routes/sudoku/getIsBoardValid.ts
type Board = Array<Array<string>>;

const EMPTY_ITEM = "_";

function getIsAreaUnique(
  board: Board,
  startColumn: number,
  startRow: number,
  ignoreEmpty: boolean
) {
  const areaValues = [];
  for (let columnIdx = startColumn; columnIdx < startColumn + 3; columnIdx++) {
    for (let rowIdx = startRow; rowIdx < startRow + 3; rowIdx++) {
      const curItem = board[rowIdx][columnIdx];
      if (curItem == EMPTY_ITEM && ignoreEmpty) continue;
      if (areaValues.includes(curItem)) return false;
      areaValues.push(curItem);
    }
  }
  return true;
}

function getIsRowUnique(board: Board, rowIdx: number, ignoreEmpty: boolean) {
  const areaValues = [];
  for (let columnIdx = 0; columnIdx < 9; columnIdx++) {
    const curItem = board[rowIdx][columnIdx];
    if (curItem == EMPTY_ITEM && ignoreEmpty) continue;
    if (areaValues.includes(curItem)) return false;
    areaValues.push(curItem);
  }
  return true;
}

function getIsColumnUnique(
  board: Board,
  columnIdx: number,
  ignoreEmpty: boolean
) {
  const areaValues = [];
  for (let rowIdx = 0; rowIdx < 9; rowIdx++) {
    const curItem = board[rowIdx][columnIdx];
    if (curItem == EMPTY_ITEM && ignoreEmpty) continue;
    if (areaValues.includes(curItem)) return false;
    areaValues.push(curItem);
  }
  return true;
}

function getIsBoardValid(board: Board, ignoreEmpty = false) {
  for (let rowIdx = 0; rowIdx < 9; rowIdx += 3) {
    for (let columnIdx = 0; columnIdx < 9; columnIdx += 3)
      if (!getIsAreaUnique(board, rowIdx, columnIdx, ignoreEmpty)) return false;
  }
  for (let rowIdx = 0; rowIdx < 9; rowIdx++) {
    if (!getIsRowUnique(board, rowIdx, ignoreEmpty)) return false;
  }

  for (let columnIdx = 0; columnIdx < 9; columnIdx++) {
    if (!getIsColumnUnique(board, columnIdx, ignoreEmpty)) return false;
  }
  return true;
}

export default getIsBoardValid;
